### DISCUSSION TITLE
Character-based CTC decoding with word constraints

### DISCUSSION THREAD

**Discussion by tshmak** [speechbrain/speechbrain#2400] [Q&A] [answered]
I'm relatively new to ASR and to speechbrain, so pardon me if the questions are stupid. 

My search for CTC decoders led me to https://speechbrain.readthedocs.io/en/latest/API/speechbrain.decoders.ctc.html 

1. I'm slightly puzzled as to why CTC with beam search and CTC with greedy decoding are presented in 2 different interfaces (1 class (CTCPrefixScorer) and 1 function (ctc_greedy_decode)). Is this deliberate? Or is it because they were contributed by different people at different times? 
2. I'm hoping to impose some kind of word constraints in the CTC decoding, but it seems this is not supported by CTCPrefixScorer. I see that decoders.seq2seq module provides other decoders which are a hybrid of CTC and some other decoders. These allow users to supply a language model also. Is this the way to implement a character-based CTC decoder with word constraints? 

Thanks in advance.

**Comment by Adel-Moumen** [comment]
Dear @tshmak, 

Thanks for your question. 

First, I'd like to say that `CTCPrefixScorer` and `ctc_greedy_decode` are two different functions as you explained. The reason why is that `CTCPrefixScorer` is a label-synchronous CTC scorer used in CTC/attention joint decoding (see: https://www.merl.com/publications/docs/TR2017-103.pdf). The CTC scorer is used as a second pass scorer on the output tokens by your attention model while  `ctc_greedy_decode` is a frame-synchronous CTC decoder that uses a greedy approach. 

You should use `CTCPrefixScorer` when you are using a CTC/attention E2E model while `ctc_greedy_decode` when you are using a pure CTC model. 

Also, I would like to invite you to have a look at https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/decoders/ctc.py , this is the new pure CTC interfaces that implements CTC beam searchers (frame-synchronous CTC). One CTC Beam searcher (https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/decoders/ctc.py#L1912) is a wrapper around torchaudio CTC decoder (https://pytorch.org/audio/main/tutorials/asr_inference_with_ctc_decoder_tutorial.html) and leave you the possibility to have constraint search. Additionally, if you want, you can also train a n-gram model on your constrain vocab, and then use it with any of our CTC decoders in  (https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/decoders/ctc.py#L1912) as it is very similar to having a constraint dict (the ngram will penalise unknown tokens/words). 

Hope I was helpful. Please let me know if you need aditionnal informations. These news decoders are still in beta but we got very good results on a variety of datasets. 

Best, 
Adel Moumen

Finally, we are working on k2 integration inside of speechbrain (https://github.com/speechbrain/speechbrain/pull/2345 should be merged very very soon) and it leaves also the possibility to the user to perform constraint search.

**Reply by tshmak** [reply, marked as answer]
Thanks Adel! That's very clear.
